### PR TITLE
Long-form teaser-thread strategy + atmosphere_long_form_composition filter

### PR DESCRIPTION
**Pull request by kraftbj**
Fixes DOTCOM-16886
See DOTCOM-16810 (parent epic)

## Proposed changes:

Introduces long-form composition strategies for Bluesky posts, with the default (`link-card`) preserved so existing sites see no change.

- New `atmosphere_long_form_composition` filter selects the long-form composition strategy. Accepts `'link-card'` (default, unchanged behavior), `'truncate-link'` (single post, body-as-text + permalink), or `'teaser-thread'` (2-post thread: hook + CTA-with-link).
- New `atmosphere_teaser_thread_posts` filter lets downstream override the default 2-post composition (e.g. expand to 3 posts, customize CTA copy).
- New `_atmosphere_bsky_thread_records` post meta: ordered array of `{ uri, cid, tid }` triples tracking every bsky post in a thread. Existing `_atmosphere_bsky_uri` / `_atmosphere_bsky_tid` / `_atmosphere_bsky_cid` continue to mirror the root post for backwards compatibility.
- New `_atmosphere_bsky_orphan_records` post meta: populated only when a thread publish fails and the compensating-delete rollback also fails, so operators have a durable manifest of records alive on the PDS but no longer tracked locally.
- New `atmosphere_pre_apply_writes` filter: short-circuits `API::apply_writes` when a non-null value is returned. Used by tests and external harnesses (FOSSE e2e) to mock/inspect the write batch.
- `Publisher::publish/update/delete` reworked to handle multi-record threads via sequential writes with partial-meta writes and compensating-delete rollback on failure.
- `Atmosphere::on_before_delete` now reads `META_THREAD_RECORDS` and schedules a delete covering every bsky tid, so force-deleting a thread-strategy post cleans up the whole thread instead of leaking the non-root replies.

Why sequential writes for thread publishes instead of one atomic `applyWrites`? Reply records need the parent's CID. CIDs are deterministic hashes of canonical DAG-CBOR — clients *can* compute them ahead of time (the TypeScript SDK does this to batch threads atomically) — but there is no DAG-CBOR encoder in our PHP dependencies yet, so we rely on the PDS to return the CID in its create response. We mitigate the partial-failure window with partial-meta writes after each successful create (so crash recovery can surface any orphans).

See [FOSSE SDD](https://github.com/Automattic/fosse/tree/trunk/sdd/long-form-bluesky-strategy) for full design context, alternatives considered, and open questions.

**Default behavior is unchanged.** `atmosphere_long_form_composition` defaults to `'link-card'`. Existing sites behave exactly as today unless they opt into a different strategy (or install a downstream like FOSSE that projects the value).

### `Publisher::update()` semantics for thread-strategy posts:

- **Same record count → in-place update.** When the new composition has the same number of records as the stored thread, every record (root + replies + doc) is rewritten via `applyWrites#update` in one atomic batch. TIDs and URIs are preserved, so external replies remain valid and Bluesky URLs stay stable. Each record gets a new CID; reply refs are built from the pre-update CIDs (self-consistent at write time, and clients treat any post-update CID mismatch as \"parent was edited\" rather than broken). \`createdAt\` is pinned to \`post_date_gmt\`.
- **Different record count or strategy change → delete + republish.** When the shape changes (link-card ↔ teaser-thread, or thread length changes), every existing record is deleted and the post is republished fresh. In this path, external replies to the original thread are orphaned (their \`reply.root\` points at a deleted AT URI), and the replaced posts surface to followers with a fresh \`createdAt\`.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

* \`composer test\` — unit tests cover composition branches, sequential-write happy path, rollback on mid-thread failure, rollback-of-rollback with orphan-manifest persistence, in-place thread update, strategy-change rewrite, delete covering the whole thread, and legacy single-record fallback.
* Manual: install with \`fosse-long-form-strategy\` downstream (or call \`add_filter( 'atmosphere_long_form_composition', fn() => 'teaser-thread' )\`), publish a titled post of 500+ words, verify two \`app.bsky.feed.post\` records land with \`reply\` refs. Edit the post content and verify the two records are updated in place (same TIDs/URIs, new content).

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Minor

#### Type

-   [x] Added - for new features

#### Message

Long-form posts can now be published as a Bluesky thread. A new filter lets sites choose between a single link-card post (default, unchanged from today), a single post that combines body text with the permalink, or a 2-post teaser thread that leads with a hook and follows with a \"continue reading\" link. Post edits update every record in place when the shape of the post hasn't changed, so Bluesky URLs stay stable and people's replies don't get orphaned. If the shape of the post changes (for example, switching between a single post and a thread), the old records are replaced with fresh ones.

</details>

### CODE REVIEW

**Comment by pfefferle** [comment]
Early review — read through the branch, a few things worth flagging before this gets much further. Ordered by impact.

### 1. Default behavior **is** changed: `createdAt` shifts from post date to publish-run time

The description says *"Default behavior is unchanged — `atmosphere_long_form_composition` defaults to `'link-card'`, existing sites behave exactly as today."* That isn't quite right.

- Today: `Post::transform()` sets `createdAt` from `$post->post_date_gmt` (line 99).
- This PR: for long-form (including the default `link-card`), the record goes through `build_long_form_records()` → `record_for_link_card()`, which **omits `createdAt`**. `publish_single` then stamps `\wp_date( 'c' )` — the current time.

Consequences:
- A normal long-form publish shows up on bsky.app with a timestamp different from the WP publish date.
- **Backfill is much worse:** every re-synced post gets stamped with the backfill-run timestamp, torpedoing the chronological ordering of the account's timeline.

Fix: propagate `$post->post_date_gmt` through `build_long_form_records` / `record_for_link_card` so the default strategy produces byte-identical output to `transform()`. Thread replies that are genuinely fresh posts can still use `wp_date('c')`.

### 2. Thread-aware force-delete is missing on the WP side

The new comment in `delete_by_tids` says *"force-deletion of thread-strategy posts is handled by the caller by reading `META_THREAD_RECORDS` pre-deletion and scheduling a separate cron per thread record."* But the only caller — `Atmosphere::on_before_delete` — hasn't been updated:

```php
$bsky_tid = \get_post_meta( $post_id, Transformer\Post::META_TID, true );
$doc_tid  = \get_post_meta( $post_id, Transformer\Document::META_TID, true );
\wp_schedule_single_event( \time(), 'atmosphere_delete_records', array( $bsky_tid, $doc_tid ) );
```

It only captures the root + doc. Force-deleting a thread-strategy post (empty-trash, `wp post delete --force`, `before_delete_post` from another plugin) leaks every non-root thread reply to Bluesky. This is a real data-consistency bug, not just a doc caveat.

### 3. Rollback's `partial_records` payload is effectively swallowed

`rollback_thread` returns a `WP_Error` with `partial_records` in the error data when rollback itself fails. But the cron closures that call `Publisher::publish` only read `get_error_code()` + `get_error_message()`. Nobody inspects `$error->get_error_data()['partial_records']`, so the orphan-record manifest is lost the moment the cron event returns. Either log the partial records alongside the error message, or persist them in post meta (e.g. `_atmosphere_bsky_orphan_records`) so an operator or recovery worker can find them later.

### 4. Thread update is delete-and-republish; the alternative isn't addressed

The PR calls out the follower-feed churn and orphaned-external-replies consequences and routes affected callers to "pin to a single-post strategy". But `applyWrites#update` on each existing thread record would do an in-place rewrite — same TIDs, same URIs, replies remain valid, `createdAt` preserved. The only cases where delete-and-republish is strictly required are record-count changes (thread length shrinks / grows) or strategy changes (thread ↔ single). For equal-length thread → thread, an N-record `applyWrites#update` batch would work.

Not a blocker, but worth a line in the description: *"We considered in-place thread updates and rejected them because X"*, or the code could prefer them when the shape matches.

### 5. Minor

- `$bsky_transformer->get_rkey();` called as a side-effect-only statement at the top of `publish()` then again inside `publish_single`/`publish_thread`. Pick one.
- `store_results()` and `mirror_thread_records_meta()` both write `Post::META_URI` / `META_TID` / `META_CID` for the root. Two code paths → easy to drift. Mirror-at-the-end only, or drop the mirroring from `store_results`.
- `record_for_thread_entry()` fires `atmosphere_transform_bsky_post` once per thread entry. The existing filter contract was one-call-per-post. Listeners that accumulate state (rate-limit counters, lint checks) will double-count. Worth a heads-up note in the filter docblock, or a new filter name for thread entries.
- The description says reply CIDs can only come from the PDS response, "a protocol constraint". It's actually a PHP-library constraint — CIDs are deterministic hashes of canonical DAG-CBOR and can be computed client-side (the TypeScript SDK does this to batch threads atomically). Just no DAG-CBOR encoder in the PHP codebase yet. Worth a one-line correction so the sequential-write choice reads as a pragmatic trade-off rather than a protocol requirement.

### Things that look good

- Keeping `transform()` byte-compatible for the short-form path and routing everything new through `build_long_form_records()` is the right shape.
- The partial-meta-per-step pattern in `publish_thread` (so crash recovery has a pointer to the root before step 2 starts) is the right discipline.
- The empty-body guard + `atmosphere_long_form_strategy_downgraded` observability hook is a nice touch.

**Comment by kraftbj** [comment]
Thanks for the careful read — all 5 addressed in 72ac474 (with extra review passes landing in follow-up commits). Walking through in the same order:

**1. `createdAt` drift.** Real regression. Fixed by centralising the fallback in `Publisher::publish_single` / `update_single` / `publish_thread`'s root to stamp `to_iso8601( $post->post_date_gmt )` instead of `wp_date( 'c' )`. Short-form `transform()` already set it; the long-form paths now match, so backfill and normal publishes both land on the WP publish date. Later passes also unified thread *replies* on the same timestamp (they were stamping `wp_date('c')` in the publish path while the update path used `post_date_gmt` — in-place edits would've silently rewound reply `createdAt` after any edit).

**2. Thread-aware force-delete.** Real bug. `Atmosphere::on_before_delete` now reads `Post::META_THREAD_RECORDS` pre-deletion and schedules a single `atmosphere_delete_records` cron with the full tid array. `Publisher::delete_by_tids` accepts `string|array`, so one `applyWrites` covers the whole thread (root + replies + doc). Legacy single-tid callers keep working; the doc-comment on `delete_by_tids` now reflects the new contract rather than apologizing for not matching it.

**3. Orphan manifest disappearing.** New post meta `Post::META_ORPHAN_RECORDS` — `rollback_thread` writes the partial state there when the compensating delete fails, in addition to `error_log` and the `WP_Error` data. Value is an append-only list (capped at 10 entries to bound meta_value growth) of `{ stamp, bsky_records, doc_rkey, original_error, rollback_error }` entries so repeated failures on the same post all survive for manual cleanup. Test coverage added.

**4. Delete-and-republish for thread updates.** Implemented in-place `applyWrites#update` — when `count($stored) === count($new_records)` and both > 1, `update_thread_in_place` rewrites every record + doc in one atomic batch. TIDs and URIs are preserved, external replies stay valid, `createdAt` is pinned to `post_date_gmt`. Reply refs are built from the *pre-update* CIDs (structurally self-consistent at write time; clients treat any post-update CID mismatch as "parent was edited"). After the call, `META_THREAD_RECORDS` is refreshed with the new CIDs from the response so subsequent updates chain from current state. `rewrite_thread` (delete + republish) is now only taken on a true shape change — strategy swap or different record count — and on failure now persists a rewrite-phase manifest to `META_ORPHAN_RECORDS` so operators have an audit trail of what was deleted.

**5. Minors.**
- Dropped the side-effect-only `get_rkey()` calls at the top of `publish()` — `publish_single`/`publish_thread` generate lazily when building writes, so the pre-call was just noise.
- `store_results()` renamed to `store_document_meta()` and reduced to owning doc meta only. `mirror_thread_records_meta()` is the single writer for `Post::META_URI / META_TID / META_CID`.
- The `atmosphere_transform_bsky_post` docblock now calls out that the filter fires per-record under `teaser-thread`, with a pointer for accumulating listeners on how to branch per-entry.
- You're right about the CID constraint — it's a PHP-ecosystem limitation (no DAG-CBOR encoder in our deps yet), not a protocol one. Updated the PR description.

Also fixed 4 failing tests on the way: one real `truncate_to_budget` bug (the `preg_replace` word-boundary fallback fired even when no whitespace existed, bypassing the hard-cap ellipsis), and three tests that didn't set `'post_excerpt' => ''` and got bitten by the WP test factory's auto-generated excerpt.

Follow-up review passes surfaced a few more things that landed as bonus fixes: reaction-sync now resolves reply URIs via a new `META_URI_INDEX` (previously a like/repost on a reply post wouldn't resolve back to the WP post); `atmosphere_pre_apply_writes` guards against malformed filter returns that would otherwise fatal on the typed return; `atmosphere_teaser_thread_posts` is sanitised and capped; and the legacy `stored_thread_records` fallback now requires a URI so a reserved-but-unpublished TID doesn't loop retry through non-existent PDS records.

CI is green across the full phpunit matrix. Flipping to Ready for Review.

**Review by pfefferle — COMMENTED** [review]
Reviewed correctness of the rollback paths, `update()` shape detection, `on_before_delete` thread cleanup, CID handling for replies, the empty-body / long-permalink downgrades, and back-compat for legacy single-record posts.

### Critical

**`Publisher::publish_thread` — mid-thread `update_document_bsky_ref` failure leaves threads half-published with no rollback** — `includes/class-publisher.php:222-225`

After the root+doc atomic write succeeds and meta is persisted, a `WP_Error` from `update_document_bsky_ref` returns early. The root bsky record and doc record are live on the PDS, but reply writes are silently abandoned and no rollback or orphan-manifest entry is written. Inconsistent with how every other partial-failure path is handled.

Either treat this call as best-effort (continue to replies, log the failure) or call `rollback_thread` on failure for symmetry.

### Major

**`atmosphere_teaser_thread_posts` — no minimum-entry guard** — `includes/transformer/class-post.php:628-649`
Filter docblock promises "at least 2 entries", but a filter returning a single string yields a 1-element array. `Publisher::publish()` then routes to `publish_single` (count===1), so the CTA reply is silently dropped without notice. Add a minimum-2 guard with `_doing_it_wrong()` before the filter becomes public API.

**Orphan manifest is write-only** — `includes/class-publisher.php:380-419`, `748-785`
`_atmosphere_bsky_orphan_records` is never read, surfaced in admin, or processed by cron. Operators would have to query post meta directly to discover orphans. Either ship a Site Health check / admin notice, or track as a follow-up issue before merge — currently neither exists.

**`update()` shape-change detection uses count comparison only** — `includes/class-publisher.php:485`
A `truncate-link` (count=1) → `teaser-thread` that empty-body-downgrades to `link-card` (count=1) takes the in-place path. Result is structurally correct, but the silent semantic strategy change is worth a comment or a guard.

### Minor

- **`build_long_form_records` long-permalink downgrade is silent** — `includes/transformer/class-post.php:434-438`. Empty-body guard fires `atmosphere_long_form_strategy_downgraded`; the long-permalink guard does not. Inconsistent observability.
- **`rollback_thread` uses single batch `applyWrites`** — `includes/class-publisher.php:320-335`. A transient PDS error means the whole 6-write batch fails and triggers the orphan manifest, even when 4 of 6 deletes would individually have succeeded. Trade-off, not a bug.
- **`test_update_sends_update_writes` has a conditional assertion that no-ops when auth blocks** — `tests/phpunit/tests/class-test-publisher.php:313-323`. Now that `atmosphere_pre_apply_writes` exists, the test should use `register_capture()` and assert unconditionally.
- **Inconsistent transformer reuse** — `includes/class-publisher.php:986` instantiates a new `Document($post)` instead of reusing the local `$doc_transformer`. Harmless but noisy.

### Positive

- `atmosphere_pre_apply_writes` short-circuit + `validate_apply_writes_response` is a clean test-mocking surface. Correctly distinguishes create/update (require uri+cid) from delete.
- `stored_thread_records` correctly handles bare-TID-without-URI fallback (failed prior publish), with a clear inline comment.
- 5-entry cap on `build_teaser_thread` is a sensible blast-radius limiter.
- `delete_by_tids` accepts `string|array` first arg with fallbacks — forward/backward compatible with cron events queued before the signature change.
- Reply refs use pre-update CIDs in `update_thread_in_place` — architecturally sound, since post-update CID drift is editorial state clients handle gracefully.
- Test coverage is comprehensive for documented scenarios: rollback happy path, rollback-of-rollback with orphan persistence, in-place updates (single + multi), shape-change rewrite, full-thread delete, legacy single-record path, `delete_by_tids` back-compat.

The half-publish bug is the only true blocker. Teaser-thread minimum-entry guard should also land before the filter is public. Orphan manifest gap is acceptable as a tracked follow-up if you want to ship soon.

File: includes/class-publisher.php [hunk 1]
```diff
@@ -61,46 +117,427 @@ public static function publish( \WP_Post $post ): array|\WP_Error {
...
+		$thread_records = array( $root_triple );
+		$created_at     = to_iso8601( $post->post_date_gmt );
+
+		self::store_document_meta( $post->ID, $root_result, $doc_transformer );
+		self::mirror_thread_records_meta( $post->ID, $thread_records );
+
+		$doc_ref_result = self::update_document_bsky_ref( $post, $doc_transformer );
+		if ( \is_wp_error( $doc_ref_result ) ) {
+			return $doc_ref_result;
+		}
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/class-publisher.php:235]: Addressed in b92fd1c — `publish_thread` now treats the doc-ref `putRecord` as best-effort: logs the failure and continues to the reply writes. The root+doc records remain published, `META_THREAD_RECORDS` ends up at the full thread length, so a subsequent edit takes the in-place update path instead of rewriting the root. Test coverage added.

File: includes/class-publisher.php [hunk 2]
```diff
@@ -16,42 +29,85 @@
...
+	 * `link-card` / `truncate-link` long-form strategies (via
+	 * `build_long_form_records()`'s single-element output).
+	 *
+	 * `createdAt` defaults to the post's `post_date_gmt` when the record
+	 * doesn't already carry one, so the Bluesky timeline mirrors the
+	 * WordPress publish date (critical for backfill — otherwise every
+	 * re-synced post would stamp with the backfill-run time and
+	 * collapse chronological order). `transform()` already sets it;
+	 * `build_long_form_records()` leaves it absent on purpose so a
+	 * single policy lives here.
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/class-publisher.php]: Addressed in b92fd1c — docblock now reads "transform() and the long-form record builders normally set it already; Publisher only backfills here when a record arrives without createdAt (for example, if a filter stripped it)." Matches the actual builder behavior.

File: tests/phpunit/tests/class-test-publisher.php [hunk 3]
```diff
@@ -224,4 +358,827 @@ public function test_delete_errors_without_tids() {
...
+
+		$result = Publisher::publish( $post );
+
+		$this->assertWPError( $result );
+		$this->assertSame( 'atmosphere_doc_ref_failed', $result->get_error_code() );
+
+		$thread_records = \get_post_meta( $post->ID, Post::META_THREAD_RECORDS, true );
+		$this->assertIsArray( $thread_records );
+		$this->assertCount( 1, $thread_records );
+		$this->assertNotEmpty( \get_post_meta( $post->ID, Document::META_URI, true ) );
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply tests/phpunit/tests/class-test-publisher.php]: Addressed in c6a86e8 — wrapped the `pre_http_request` filter in try/finally with a referenced callback (`$put_record_failure`) so it is removed even if assertions throw. Also reworked `test_update_sends_update_writes` to use `register_capture()` (`atmosphere_pre_apply_writes`) instead of `pre_http_request`, dropping the conditional assertions that no-opped when DPoP blocked the request.

File: includes/class-api.php [hunk 4]
```diff
@@ -167,6 +167,40 @@ public static function upload_blob( string $file_path, string $mime_type ): arra
...
+		 * before any of that.
+		 *
+		 * @param null|array|\WP_Error $short_circuit Short-circuit value. Return null to skip.
+		 * @param array                $writes        The write batch about to be sent.
+		 */
+		$short_circuit = \apply_filters( 'atmosphere_pre_apply_writes', null, $writes );
+
+		if ( \is_array( $short_circuit ) || \is_wp_error( $short_circuit ) ) {
+			return $short_circuit;
+		}
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/class-api.php:193]: Already resolved in 0e1e9cc — `apply_writes()` now routes any array short-circuit through `validate_apply_writes_response()`, which checks: `results` key present and is a list, count matches the writes batch, every entry is an array, and create/update entries have non-empty `uri` + `cid`. A non-array, non-WP_Error return also short-circuits to `WP_Error("atmosphere_invalid_pre_apply_writes_return")`. Test coverage in `test_pre_apply_writes_*`.

File: includes/transformer/class-post.php [hunk 5]
```diff
@@ -271,4 +324,351 @@ private function is_short_form( \WP_Post $post ): bool {
...
+	 * @return string
+	 */
+	private function build_truncate_link_text(): string {
+		$permalink = \get_permalink( $this->object );
+		$plain     = $this->render_post_content_plain( $this->object );
+		$budget    = 300 - \mb_strlen( "\n\n" ) - \mb_strlen( $permalink );
+
+		$body = $this->truncate_to_budget( $plain, $budget, false );
+
+		return $body . "\n\n" . $permalink;
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/transformer/class-post.php]: Already resolved in 0e1e9cc — the budget can never reach `truncate_to_budget()` non-positively. `build_truncate_link_text()` early-returns the truncated permalink alone when `mb_strlen($permalink) >= $max_length`, and short-circuits to `return $permalink` when `$budget <= mb_strlen($separator)`. After both guards, `$budget - mb_strlen($separator) >= 1`. `truncate_to_budget()` itself also defends against `$max <= 0` (line 499) and `$max === 1` (returns `…`).

**Comment by kraftbj** [comment]
@pfefferle thanks for the thorough re-read. Walking through everything in the review body:

### Critical

**Doc-ref half-publish (`publish_thread`).** Fixed in b92fd1c. The doc-ref `putRecord` is now best-effort — failure logs and continues to the reply writes, so `META_THREAD_RECORDS` ends up at the full thread length and a subsequent edit takes the in-place update path instead of rewriting the already-published root. Test coverage in `test_publish_thread_continues_when_doc_ref_update_fails`.

### Major

1. **`atmosphere_teaser_thread_posts` minimum-entry guard.** Added in c464fda. A filter return with fewer than 2 valid string entries triggers `_doing_it_wrong( 'atmosphere_teaser_thread_posts', ..., 'unreleased' )` and falls back to the default hook + CTA pair. Test in `test_build_long_form_records_teaser_thread_filter_under_two_falls_back`.

2. **Orphan manifest write-only.** Tracked as issue 44 — covers Site Health surfacing, admin notice, WP-CLI, and a retry-with-backoff cron. Added a TODO referencing it from `persist_orphan_records()` in 14f21d2.

3. **`update()` shape-change count-only detection.** Added a clarifying comment in 14f21d2 — the output is structurally correct because a `truncate-link` (count=1) → `teaser-thread` that empty-body-downgrades to `link-card` (count=1) both produce a single-record post; URIs/TIDs reuse on the bsky side. No behavior change.

### Minor

- **Long-permalink downgrade observability.** Fixed in b92fd1c — `atmosphere_long_form_strategy_downgraded` now fires from both branches in `build_long_form_records()`.
- **`rollback_thread` single batch.** Acknowledged as a trade-off; no change.
- **`test_update_sends_update_writes` conditional assertion.** Fixed in c6a86e8 — now uses `register_capture()` (`atmosphere_pre_apply_writes`) and asserts unconditionally.
- **Inconsistent transformer reuse.** Fixed in 14f21d2 — `update_document_bsky_ref()` reuses the existing `$doc_transformer` (its `transform()` reads URI/CID fresh from post meta on every call).

**Comment by kraftbj** [comment]
Codex adversarial pass surfaced three more findings; all addressed:

### Findings + fixes

1. **[HIGH] Ambiguous reply-write failures could leave untracked live replies.** `publish_thread`'s reply-create error path only rolled back already-stored thread records, so a transport-level WP_Error after a server-side commit (response timeout, connection drop) left the just-attempted reply alive on Bluesky with no WP-side handle on it. Fixed in 880af50: the locally-generated `$reply_rkey` is now packed into a synthetic triple and appended to the rollback batch. If the PDS never committed, the compensating delete is a no-op (or surfaces in the orphan manifest); if it did, rollback cleans it up. Existing rollback tests updated for the new shape.

2. **[MEDIUM] Best-effort doc-ref failures were silent permanent partial publishes.** The b92fd1c best-effort change made `error_log` the only signal — backfill callers would mark the post synced and never retry the missing `bskyPostRef`. Fixed in 694f3c5: persist `Post::META_DOC_REF_PENDING` (`{ stamp, code, message }`) when the doc-ref `putRecord` fails, clear it on the next successful `update_document_bsky_ref`. Re-saving the post is the natural recovery path; `clear_all_record_meta` clears the marker on rollback / delete. TODO comment broadens issue 44's surfacing scope to include this.

3. **[MEDIUM] Long-permalink guard missed the CTA prefix overhead.** `requires_link_card_for_long_permalink()` only checked the bare permalink (≥ 300). The actual CTA is `Continue reading: <permalink>`, so a permalink ~290 chars passed the guard and then word-cut the over-budget URL fragment off, shipping a thread whose final post said `Continue reading:` with no link. Fixed in 8374793: new `requires_link_card_for_teaser_thread()` builds the localized CTA via a `teaser_thread_cta_text()` helper and compares its `mb_strlen` against 300; `build_teaser_thread()` reuses the same helper so the guard and the actual composition can't desync across locales. truncate-link still uses the bare-permalink check (its overflow equation differs).

CI green across PHP 8.2/8.3/8.5 × WP 6.2/latest/trunk.

**Comment by pfefferle** [comment]
I hope I will find some time to review this biggie tomorrow!

**Review by pfefferle — APPROVED** [review]
Works like a charm! Do you plan to have a "real" thread with the full content some day?
